staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -406,14 +406,36 @@ func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
 	return nil
 }
 
+func (dm *discoveryManager) markInitialStale() {
+	dm.servicesLock.Lock()
+	defer dm.servicesLock.Unlock()
+
+	for key, info := range dm.apiServices {
+		gv := helper.APIServiceNameToGroupVersion(key)
+		entry := apidiscoveryv2beta1.APIVersionDiscovery{
+			Freshness: apidiscoveryv2beta1.DiscoveryFreshnessStale,
+			Version:   gv.Version,
+		}
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+		dm.mergedDiscoveryHandler.SetGroupVersionPriority(metav1.GroupVersion(gv), info.groupPriority, info.versionPriority)
+	}
+}
+
 // Spwans a goroutune which waits for added/updated apiservices and updates
 // the discovery document accordingly
-func (dm *discoveryManager) Run(stopCh <-chan struct{}) {
+func (dm *discoveryManager) Run(stopCh <-chan struct{}, discoverySyncedCh chan<- struct{}) {
 	klog.Info("Starting ResourceDiscoveryManager")
 
 	// Shutdown the queue since stopCh was signalled
 	defer dm.dirtyAPIServiceQueue.ShutDown()
 
+	// Ensure that apiregistration.k8s.io is the first group in the discovery group.
+	dm.mergedDiscoveryHandler.WithSource(discoveryendpoint.BuiltinSource).SetGroupVersionPriority(APIRegistrationGroupVersion, APIRegistrationGroupPriority, 0)
+
+	// Marks the initial list of APIServices as Stale such that they're published for before readiness check succeeds
+	dm.markInitialStale()
```

Comment:
Why take this approach instead of making a single synchronous pass through our "standard" `syncAPIService` which would either populate the initial or mark as stale?  Isn't it more useful for us to try to populate the discovery data since we know its extremely likely that someone will request it and avoid a different path adding Groups?

Reply:
syncAPIService is blocking. Aggregated apiservers could take a long time to respond and we don't want to block readiness check of a cluster on the responsiveness of an aggegated apiserver. It's also not ideal to omit the aggregated apiserver from discovery, so populating the apiservice and marking it as stale is what we decided on. Open to feedback/suggestions

Reply:
> syncAPIService is blocking. Aggregated apiservers could take a long time to respond and we don't want to block readiness check of a cluster on the responsiveness of an aggegated apiserver. It's also not ideal to omit the aggregated apiserver from discovery, so populating the apiservice and marking it as stale is what we decided on. Open to feedback/suggestions

time limited context + go func per apiservice (ending with just assigning stale on timeout) + wait-group ?